libp2p/NodeTable.cpp

```diff
@@ -20,45 +20,70 @@ constexpr unsigned c_handleTimeoutsIntervalMs = 5000;
 
 }  // namespace
 
-constexpr std::chrono::seconds DiscoveryDatagram::c_timeToLive;
-constexpr std::chrono::milliseconds NodeTable::c_reqTimeout;
-constexpr std::chrono::milliseconds NodeTable::c_bucketRefresh;
-constexpr std::chrono::milliseconds NodeTable::c_evictionCheckInterval;
+constexpr chrono::seconds DiscoveryDatagram::c_timeToLive;
+constexpr chrono::milliseconds NodeTable::c_reqTimeout;
+constexpr chrono::milliseconds NodeTable::c_bucketRefresh;
+constexpr chrono::milliseconds NodeTable::c_evictionCheckInterval;
 
-inline bool operator==(
-    std::weak_ptr<NodeEntry> const& _weak, std::shared_ptr<NodeEntry> const& _shared)
+inline bool operator==(weak_ptr<NodeEntry> const& _weak, shared_ptr<NodeEntry> const& _shared)
 {
     return !_weak.owner_before(_shared) && !_shared.owner_before(_weak);
 }
 
 NodeTable::NodeTable(ba::io_service& _io, KeyPair const& _alias, NodeIPEndpoint const& _endpoint,
     bool _enabled, bool _allowLocalDiscovery)
-  : m_hostNodeID(_alias.pub()),
-    m_hostNodeEndpoint(_endpoint),
-    m_secret(_alias.secret()),
-    m_socket(make_shared<NodeSocket>(
-        _io, static_cast<UDPSocketEvents&>(*this), (bi::udp::endpoint)m_hostNodeEndpoint)),
-    m_requestTimeToLive(DiscoveryDatagram::c_timeToLive),
-    m_allowLocalDiscovery(_allowLocalDiscovery),
-    m_timers(_io),
-    m_io(_io)
+  : m_hostNodeID{_alias.pub()},
+    m_hostNodeEndpoint{_endpoint},
+    m_secret{_alias.secret()},
+    m_socket{make_shared<NodeSocket>(
+        _io, static_cast<UDPSocketEvents&>(*this), (bi::udp::endpoint)m_hostNodeEndpoint)},
+    m_requestTimeToLive{DiscoveryDatagram::c_timeToLive},
+    m_allowLocalDiscovery{_allowLocalDiscovery},
+    m_discoveryTimer{make_shared<ba::deadline_timer>(_io)},
+    m_evictionTimer{make_shared<ba::deadline_timer>(_io)},
+    m_io{_io}
 {
     for (unsigned i = 0; i < s_bins; i++)
         m_buckets[i].distance = i;
 
     if (!_enabled)
+    {
+        cwarn << "\"_enabled\" parameter is false, discovery is disabled";
         return;
+    }
 
     try
     {
         m_socket->connect();
-        doDiscovery();
-        doHandleTimeouts();
     }
-    catch (std::exception const& _e)
+    catch (exception const& _e)
     {
         cwarn << "Exception connecting NodeTable socket: " << _e.what();
         cwarn << "Discovery disabled.";
+        return;
+    }
+}
+
+void NodeTable::start()
+{
+    if (!m_socket->isOpen())
+    {
+        cwarn << "NodeTable socket isn't open. Discovery is disabled";
+        return;
+    }
+
+    try
+    {
+        doDiscovery();
+        m_evictionTimer->expires_from_now(
+            boost::posix_time::milliseconds(c_handleTimeoutsIntervalMs));
+        m_evictionTimer->async_wait(bind(
```

Comment:
Better use lambdas instead of `bind`.
Lambda syntax is more common, and also `bind` can have some obscure problems

Reply:
Ah ok, went with bind since the boost examples i looked at used it and the syntax is a little more intuitive to me. For lambda should we just capture the shared_from_this by value?